src/Selenium2Driver.php

```diff
@@ -758,170 +695,135 @@ public function selectOption($xpath, $value, $multiple = false)
 
     /**
      * {@inheritdoc}
      */
     public function rightClick($xpath)
     {
-        $this->mouseOver($xpath);
-        $this->wdSession->click(array('button' => 2));
+        $element = $this->findElement($xpath);
+        $this->webDriver->action()->contextClick($element)->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function attachFile($xpath, $path)
     {
+        /** @var RemoteWebElement $element */
         $element = $this->findElement($xpath);
         $this->ensureInputType($element, $xpath, 'file', 'attach a file on');
 
-        // Upload the file to Selenium and use the remote path. This will
-        // ensure that Selenium always has access to the file, even if it runs
-        // as a remote instance.
-        try {
-          $remotePath = $this->uploadFile($path);
-        } catch (\Exception $e) {
-          // File could not be uploaded to remote instance. Use the local path.
-          $remotePath = $path;
-        }
-
-        $element->postValue(array('value' => array($remotePath)));
+        $element->setFileDetector(new LocalFileDetector());
+        return $element->sendKeys($path);
     }
 
     /**
      * {@inheritdoc}
      */
     public function isVisible($xpath)
     {
-        return $this->findElement($xpath)->displayed();
+        $element = $this->findElement($xpath);
+        return $element->isDisplayed();
     }
 
     /**
      * {@inheritdoc}
      */
     public function mouseOver($xpath)
     {
-        $this->wdSession->moveto(array(
-            'element' => $this->findElement($xpath)->getID()
-        ));
+        $element = $this->findElement($xpath);
+        $this->webDriver->action()->moveToElement($element)->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function focus($xpath)
     {
-        $this->trigger($xpath, 'focus');
+        $element = $this->findElement($xpath);
+        $action = $this->webDriver->action();
+
+        $action->moveToElement($element);
+        $action->click($element);
+        $action->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function blur($xpath)
     {
-        $this->trigger($xpath, 'blur');
+        $element = $this->findElement($xpath);
+        $action = $this->webDriver->action();
+
+        $action->moveToElement($element);
+        $action->sendKeys($element, WebDriverKeys::TAB);
+        $action->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function keyPress($xpath, $char, $modifier = null)
     {
-        $options = self::charToOptions($char, $modifier);
-        $this->trigger($xpath, 'keypress', $options);
+        $this->sendKey($xpath, $char, $modifier);
     }
 
     /**
      * {@inheritdoc}
      */
     public function keyDown($xpath, $char, $modifier = null)
     {
-        $options = self::charToOptions($char, $modifier);
-        $this->trigger($xpath, 'keydown', $options);
+        $this->sendKey($xpath, $char, $modifier);
```

Comment:
that's not the same API than previously. After calling keyDown, your implementation already released the key.

I think we should rather add a new `pressKey` method in Mink, and deprecate these 3 methods. And the new version of the driver would then implement only `pressKey`.
And we should implement the new method in 1.x too

The old implementation was not good anyway, as it was not actually pressing the key, but only triggering the JS actions (so the default behavior of pressing a key would not happen).

Reply:
https://github.com/minkphp/Mink/pull/772